File: src/services/location.ts
import ApiService from './api';

export interface LocationData {
  latitude: number;
  longitude: number;
  address: string;
}

export interface PlaceData {
  id: string;
  name: string;
  address: string;
  latitude: number;
  longitude: number;
  category: string;
  relevance: number;
  distance?: number;
}

class LocationService {
  private static instance: LocationService;
  private currentLocation: LocationData | null = null;

  static getInstance(): LocationService {
    if (!LocationService.instance) {
      LocationService.instance = new LocationService();
    }
    return LocationService.instance;
  }

  async getCurrentLocation(): Promise<LocationData> {
    return new Promise((resolve, reject) => {
      if (!navigator.geolocation) {
        reject(new Error('Geolocation is not supported by this browser'));
        return;
      }

      navigator.geolocation.getCurrentPosition(
        async (position) => {
          const { latitude, longitude } = position.coords;
          
          try {
            // Get address from coordinates using our backend
            const response = await ApiService.request('/api/v1/location/current-location', {
              method: 'POST',
              headers: {
                'Content-Type': 'application/json',
              },
              body: JSON.stringify({
                latitude,
                longitude
              }),
            });

            this.currentLocation = {
              latitude,
              longitude,
              address: response.address
            };

            resolve(this.currentLocation);
          } catch (error) {
            console.error('Error getting address from coordinates:', error);
            // Fallback to coordinates if address lookup fails
            this.currentLocation = {
              latitude,
              longitude,
              address: `${latitude.toFixed(4)}, ${longitude.toFixed(4)}`
            };
            resolve(this.currentLocation);
          }
        },
        (error) => {
          console.error('Error getting location:', error);
          reject(new Error('Unable to retrieve your location'));
        },
        {
          enableHighAccuracy: true,
          timeout: 10000,
          maximumAge: 300000 // 5 minutes
        }
      );
    });
  }

  async searchPlaces(query: string, currentLocation?: LocationData): Promise<PlaceData[]> {
    try {
      const response = await ApiService.request('/api/v1/location/search-places', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          query,
          latitude: currentLocation?.latitude,
          longitude: currentLocation?.longitude
        }),
      });

      return response;
    } catch (error) {
      console.error('Error searching places:', error);
      return [];
    }
  }

  async getNearbyPlaces(currentLocation: LocationData): Promise<PlaceData[]> {
    try {
      const response = await ApiService.request('/api/v1/location/nearby-places', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          latitude: currentLocation.latitude,
          longitude: currentLocation.longitude
        }),
      });

      return response;
    } catch (error) {
      console.error('Error getting nearby places:', error);
      return [];
    }
  }

  getStoredLocation(): LocationData | null {
    return this.currentLocation;
  }

  clearLocation(): void {
    this.currentLocation = null;
  }
}

export default LocationService.getInstance();

File: rideshare/app/api/v1/endpoints/location.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import List, Optional
from app.services.location_service import location_service

router = APIRouter()

class LocationRequest(BaseModel):
    latitude: float
    longitude: float

class SearchPlacesRequest(BaseModel):
    query: str
    latitude: Optional[float] = None
    longitude: Optional[float] = None

class LocationResponse(BaseModel):
    address: str
    latitude: float
    longitude: float
    context: List[dict] = []

class PlaceResponse(BaseModel):
    id: str
    name: str
    address: str
    latitude: float
    longitude: float
    category: str = ""
    relevance: float = 0.0
    distance: Optional[float] = None

@router.post("/current-location", response_model=LocationResponse)
async def get_current_location(request: LocationRequest):
    """Get current location details from coordinates"""
    try:
        location = await location_service.get_current_location(
            request.latitude, 
            request.longitude
        )
        
        if not location:
            raise HTTPException(
                status_code=404,
                detail="Location not found"
            )
        
        return LocationResponse(**location)
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to get location: {str(e)}"
        )

@router.post("/search-places", response_model=List[PlaceResponse])
async def search_places(request: SearchPlacesRequest):
    """Search for places using query"""
    try:
        places = await location_service.search_places(
            request.query,
            request.latitude,
            request.longitude
        )
        
        return [PlaceResponse(**place) for place in places]
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to search places: {str(e)}"
        )

@router.post("/nearby-places", response_model=List[PlaceResponse])
async def get_nearby_places(request: LocationRequest):
    """Get nearby places from current location"""
    try:
        places = await location_service.get_nearby_places(
            request.latitude,
            request.longitude
        )
        
        return [PlaceResponse(**place) for place in places]
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to get nearby places: {str(e)}"
        )